pandas/core/arrays/categorical.py

```diff
@@ -619,15 +619,43 @@ def from_codes(cls, codes, categories, ordered=False):
         codes : array-like, integers
             An integer array, where each integer points to a category in
             categories or -1 for NaN
-        categories : index-like
+        categories : index-like, optional
             The categories for the categorical. Items need to be unique.
-        ordered : boolean, (default False)
-            Whether or not this categorical is treated as a ordered
+
+            .. deprecated:: 0.24.0
+                Use ``dtype`` instead.
+        ordered : bool, optional
+            Whether or not this categorical is treated as an ordered
             categorical. If not given, the resulting categorical will be
             unordered.
-        """
-        dtype = CategoricalDtype._from_values_or_dtype(codes, categories,
-                                                       ordered)
+
+            .. deprecated:: 0.24.0
+                Use ``dtype`` instead.
+        dtype : CategoricalDtype
+            An instance of ``CategoricalDtype`` to use for this categorical.
+
+            .. versionadded:: 0.24.0
+                dtype will be required in the future.
+
+        Examples
+        --------
+        >>> dtype = pd.api.types.CategoricalDtype(['a', 'b'], ordered=True)
+        >>> pd.Categorical.from_codes(codes=[0, 1, 0, 1], dtype=dtype)
+        [a, b, a, b]
+        Categories (2, object): [a < b]
+        """
+        if dtype is not None:
+            if categories is not None or ordered is not None:
+                raise ValueError("Cannot specify `categories` or `ordered` "
+                                 "together with `dtype`.")
+        elif categories is None and dtype is None:
+            raise ValueError("Must specify `dtype`.")
+        else:
+            msg = u("The 'categories' and 'ordered' keyword are deprecated "
```

Comment:
Passing `categories=...` is pretty convenient and widely used. I'm not sure they should be deprecated here.